### PR TITLE
Add SSN model with hashing and salting

### PR DESCRIPTION
**Pull request by macrael**
## Description

This adds a SocialSecurityNumber model to our models. It stores SSN's hashed with bcrypt so that if the db were compromised none of the SSNs could be recovered. Note that while no salt field is present in the database, the go bcrypt library embeds the salt into the returned hash (and then uses that salt when verifying) so it is indeed salted. A test verifies this.

This PR then hooks up that to the existing service member models for create and patch. 

## Reviewer Notes

For now I put the foreign key on the SSN itself, but I think it might be more correct to put it on the User model to enforce that each User has at most one SSN associated with them. I'm curious about your thoughts. 

Also, this is code that deals with encryption, so please be make sure we aren't doing anything that would make SSNs be recoverable. 

## Verification Steps

* [ ] All tests pass.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/blob/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.

## References

* https://www.pivotaltracker.com/story/show/156442165
* https://godoc.org/golang.org/x/crypto/bcrypt

### CODE REVIEW

File: pkg/models/social_security_number.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	ID            uuid.UUID `db:"id"`
+	CreatedAt     time.Time `db:"created_at"`
+	UpdatedAt     time.Time `db:"updated_at"`
+	UserID        uuid.UUID `db:"user_id"`
+	EncryptedHash string    `db:"encrypted_hash"`
+}
+
+// SocialSecurityNumbers is a list of SSNs
+type SocialSecurityNumbers []SocialSecurityNumber
+
```
Comment (ntwyman) [inline pkg/models/social_security_number.go]: I think something in here should validate that the SSN is properly formatted. I am concerned that an ssn sneaks through with, e.g. 123_12-1234 and then we can never match on it again. A simple \d{3}-\d{2}-\d{4} pattern match would be enough

File: pkg/models/social_security_number.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	hash := string(byteHash)
+	ssn := SocialSecurityNumber{
+		UserID:        userID,
+		EncryptedHash: hash,
+	}
+	return ssn, nil
+}
+
+// MatchesRawSSN returns true if the encrypted_hahs matches the unencryptedSSN
+func (s SocialSecurityNumber) MatchesRawSSN(unencryptedSSN string) bool {
```
Comment (ntwyman) [inline pkg/models/social_security_number.go]: I think Matches is good enough for this. It's contextually clear.

File: pkg/models/social_security_number.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	return validate.NewErrors(), nil
+}
+
+// ErrSSNBadFormat is returned if you attempted to hash an SSN not in the format '123-12-1234'
+var ErrSSNBadFormat = errors.New("SSNs must be in the format '123-12-1234'")
+var ssnFormatValidator = regexp.MustCompile(`^\d{3}-\d{2}-\d{4}$`)
+
+// BuildSocialSecurityNumber returns an *unsaved* SSN that has the ssn hash set based on the passed in raw ssn
+func BuildSocialSecurityNumber(userID uuid.UUID, unencryptedSSN string) (SocialSecurityNumber, error) {
+	if !ssnFormatValidator.Match([]byte(unencryptedSSN)) {
```
[the remark this answers is not in the document]
Comment (macrael) [reply pkg/models/social_security_number.go]: I like that better, missed that method

**Comment by macrael** [comment]
Note, this is not ready to merge, it ended up a bit out of order in the queue, it will be merged once the ServiceMember handlers have been created.

**Pull request merged**